pkg/kubelet/kubelet.go

```diff
@@ -1635,6 +1646,59 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
 	return nil
 }
 
+func (kl *Kubelet) canResizePod(pod *v1.Pod) (bool, string) {
+	var otherActivePods []*v1.Pod
+	kl.podResizeMutex.Lock()
+	defer kl.podResizeMutex.Unlock()
+	activePods := kl.GetActivePods()
+	for _, p := range activePods {
+		if p.UID != pod.UID {
+			otherActivePods = append(otherActivePods, p)
+		}
+	}
+	if ok, failReason, failMessage := kl.canAdmitPod(otherActivePods, pod); !ok {
+		// Log reason and return. Let the next sync iteration retry the resize
+		klog.V(2).Infof("Pod '%s' resize cannot be accommodated. Reason: '%s' Message: '%s'", pod.Name, failReason, failMessage)
+		return false, ""
+	}
+
+	var containersPatchData string
+	for _, container := range pod.Spec.Containers {
+		var resourcesPatchData string
+		for rName, rQuantity := range container.Resources.Requests {
+			container.ResourcesAllocated[rName] = rQuantity
+			resourcesPatchData += fmt.Sprintf(`"%s":"%s",`, rName, rQuantity.String())
+		}
+		resourcesPatchData = strings.TrimRight(resourcesPatchData, ",")
+		containersPatchData += fmt.Sprintf(`{"name":"%s","resourcesAllocated":{%s}},`, container.Name, resourcesPatchData)
+	}
+	containersPatchData = strings.TrimRight(containersPatchData, ",")
+	patchData := fmt.Sprintf(`{"spec":{"containers":[%s]}}`, containersPatchData)
+	return true, patchData
```

Comment:
We have a race condition here, but i don't remember how we agreed to solve it.  For example, if we called canResizePod on pod A and B, such that the ordering was:

1. canResizePod(A) => true
2. canResizePod(B) => true
3. Patch(A)
4. Patch(B)

This could pass even if there was only room for 1 of the pods, since A is not included in kl.GetActivePods() during canResizePod(B).

At minimum, I think we need to update GetActivePods here.  Or am I misremembering our conclusion?

Reply:
I've tested this for add vs resize race after coding up the switch-over from Requests to ResourcesAllocated (next change - hopefully by eod Monday). I think the mutex should serialize race between two resizes as well.

So in the above case, the next GetActivePods (for pod B( should use updated ResourcesAllocated values of Pod A if the resize was accepted. But let me specifically test that case and verify..

Reply:
I tested the case of concurrent updates, and the mutex mechanism holds up well.

Reply:
Where do we update the active pods?

Reply:
The other thing I was trying to reason through at sig-node is whether the update needs to be written back to the API Server while holding the lock.  Take the following example:

1. canResizePod(A) decrease by 100m => true
2. updateActivePods(A) decrease by 100m
3. canResizePod(B) increase by 100m => true, since A's requests are lowered
4. updateActivePods(B) increase by 100m
5. Patch(B)
6. Update (B) in the CRI
6. ... Patch(A) hangs for a while ... Update(A) in the CRI doesn't happen.

In this case, B would have a higher allocated even though A has not been patched to be lower.  An observer that did `kubectl describe no` would see resource allocated > allocatable.  We could decide that that is OK, since we _really_ just care about changes sent to the CRI, and the Patch() doesn't change that.

Does that make sense?

Reply:
> Where do we update the active pods?

In canResizePod function after successful admit of resized pod when we assign ```container.ResourcesAllocated[rName] = rQuantity```

Here's what it looks like when we slow it down (below code) and play it (log down below) . 16 cpu allocatable, pod1 resizes first from 2 cpu to 10 cpu, pod2 resizes from 3 cpu to 6 cpu, but does not get it - active pods has been updated when pod2 pulls up active pods even though pod1 apiserver patch happens much later.

```
diff --git a/pkg/kubelet/kubelet.go b/pkg/kubelet/kubelet.go
index 7c0c0cc..ecf9295 100644
--- a/pkg/kubelet/kubelet.go
+++ b/pkg/kubelet/kubelet.go
@@ -122,6 +122,7 @@ import (
        "k8s.io/kubernetes/pkg/volume/util/hostutil"
        "k8s.io/kubernetes/pkg/volume/util/subpath"
        "k8s.io/kubernetes/pkg/volume/util/volumepathhandler"
+"k8s.io/kubernetes/pkg/api/v1/resource"
 )
 
 const (
@@ -1648,10 +1649,12 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
 
 func (kl *Kubelet) canResizePod(pod *v1.Pod) (bool, string) {
        var otherActivePods []*v1.Pod
+klog.Infof(">>__DBG_canResizePod_ENTER: pod=%s. Allocations: %+v\n", pod.Name, resource.PodResourceAllocations(pod))
        kl.podResizeMutex.Lock()
        defer kl.podResizeMutex.Unlock()
        activePods := kl.GetActivePods()
        for _, p := range activePods {
+klog.Infof(">>__DBG_canResizePod_ACTIVE_POD: pod=%s. active_pod=%s, Allocations: %+v\n", pod.Name, p.Name, resource.PodResourceAllocations(p))
                if p.UID != pod.UID {
                        otherActivePods = append(otherActivePods, p)
                }
@@ -1659,6 +1662,7 @@ func (kl *Kubelet) canResizePod(pod *v1.Pod) (bool, string) {
        if ok, failReason, failMessage := kl.canAdmitPod(otherActivePods, pod); !ok {
                // Log reason and return. Let the next sync iteration retry the resize
                klog.V(2).Infof("Pod '%s' resize cannot be accommodated. Reason: '%s' Message: '%s'", pod.Name, failReason, failMessage)
+klog.Infof(">>__DBG_canResizePod_RESIZE_ADMIT_FAILED: pod=%s. reason: %s, message: %s\n", pod.Name, failReason, failMessage)
                return false, ""
        }
 
@@ -1672,6 +1676,8 @@ func (kl *Kubelet) canResizePod(pod *v1.Pod) (bool, string) {
                resourcesPatchData = strings.TrimRight(resourcesPatchData, ",")
                containersPatchData += fmt.Sprintf(`{"name":"%s","resourcesAllocated":{%s}},`, container.Name, resourcesPatchData)
        }
+klog.Infof(">>__DBG_canResizePod_RESIZE_PATCH_CREATED: pod=%s. Allocations: %+v\n", pod.Name, resource.PodResourceAllocations(pod))
+time.Sleep(30 * time.Second) // CONCURRENCY_CREATOR
        containersPatchData = strings.TrimRight(containersPatchData, ",")
        patchData := fmt.Sprintf(`{"spec":{"containers":[%s]}}`, containersPatchData)
        return true, patchData
@@ -1690,10 +1696,13 @@ func (kl *Kubelet) handlePodResourcesResize(pod *v1.Pod) {
        }
 
        if fit, patchData := kl.canResizePod(pod); fit {
+klog.Infof(">>__DBG_handlePodResourcesResize_PATCHING_APISERVER: pod=%s\n", pod.Name)
+if pod.Name == "pod1" { time.Sleep(30 * time.Second) } // APISERVER_BUSY_EMULATOR
                _, patchError := kl.kubeClient.CoreV1().Pods(pod.Namespace).Patch(context.TODO(), pod.Name, types.StrategicMergePatchType, []byte(patchData), metav1.PatchOptions{})
                if patchError != nil {
                        klog.Errorf("Failed to patch ResourcesAllocated values for pod %s: %+v\n", pod.Name, patchError)
                }
+klog.Infof(">>__DBG_handlePodResourcesResize_PATCHING_APISERVER_DONE: pod=%s\n", pod.Name)
        }
        return
```
---------

```
root@fw0000358:~/jun01_vkInPlacePodVS_v1_19_dev/pkg/kubelet# cat /tmp/kubelet.log | grep __DBG_
I0609 21:44:02.198922   30286 kubelet.go:1652] >>__DBG_canResizePod_ENTER: pod=pod1. Allocations: map[cpu:{i:{value:2 scale:0} d:{Dec:<nil>} s:2 Format:DecimalSI}]
I0609 21:44:02.198966   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod1. active_pod=kube-dns-747b6894bc-h8d95, Allocations: map[cpu:{i:{value:260 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} memory:{i:{value:115343360 scale:0} d:{Dec:<nil>} s: Format:BinarySI}]
I0609 21:44:02.198999   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod1. active_pod=pod1, Allocations: map[cpu:{i:{value:2 scale:0} d:{Dec:<nil>} s:2 Format:DecimalSI}]
I0609 21:44:02.199025   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod1. active_pod=pod2, Allocations: map[cpu:{i:{value:3 scale:0} d:{Dec:<nil>} s:3 Format:DecimalSI}]
I0609 21:44:02.200486   30286 kubelet.go:1679] >>__DBG_canResizePod_RESIZE_PATCH_CREATED: pod=pod1. Allocations: map[cpu:{i:{value:10 scale:0} d:{Dec:<nil>} s:10 Format:DecimalSI}]
I0609 21:44:09.774348   30286 kubelet.go:1652] >>__DBG_canResizePod_ENTER: pod=pod2. Allocations: map[cpu:{i:{value:3 scale:0} d:{Dec:<nil>} s:3 Format:DecimalSI}]
I0609 21:44:32.200830   30286 kubelet.go:1699] >>__DBG_handlePodResourcesResize_PATCHING_APISERVER: pod=pod1
I0609 21:44:32.200889   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod2. active_pod=pod2, Allocations: map[cpu:{i:{value:3 scale:0} d:{Dec:<nil>} s:3 Format:DecimalSI}]
I0609 21:44:32.200945   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod2. active_pod=kube-dns-747b6894bc-h8d95, Allocations: map[cpu:{i:{value:260 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} memory:{i:{value:115343360 scale:0} d:{Dec:<nil>} s: Format:BinarySI}]
I0609 21:44:32.200986   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod2. active_pod=pod1, Allocations: map[cpu:{i:{value:10 scale:0} d:{Dec:<nil>} s:10 Format:DecimalSI}]
I0609 21:44:32.202919   30286 kubelet.go:1665] >>__DBG_canResizePod_RESIZE_ADMIT_FAILED: pod=pod2. reason: OutOfcpu, message: Node didn't have enough resource: cpu, requested: 6000, used: 10260, capacity: 16000
I0609 21:44:32.831605   30286 kubelet.go:1652] >>__DBG_canResizePod_ENTER: pod=pod2. Allocations: map[cpu:{i:{value:3 scale:0} d:{Dec:<nil>} s:3 Format:DecimalSI}]
I0609 21:44:32.831646   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod2. active_pod=pod2, Allocations: map[cpu:{i:{value:3 scale:0} d:{Dec:<nil>} s:3 Format:DecimalSI}]
I0609 21:44:32.831674   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod2. active_pod=kube-dns-747b6894bc-h8d95, Allocations: map[cpu:{i:{value:260 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} memory:{i:{value:115343360 scale:0} d:{Dec:<nil>} s: Format:BinarySI}]
I0609 21:44:32.831703   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod2. active_pod=pod1, Allocations: map[cpu:{i:{value:10 scale:0} d:{Dec:<nil>} s:10 Format:DecimalSI}]
I0609 21:44:32.833177   30286 kubelet.go:1665] >>__DBG_canResizePod_RESIZE_ADMIT_FAILED: pod=pod2. reason: OutOfcpu, message: Node didn't have enough resource: cpu, requested: 6000, used: 10260, capacity: 16000
I0609 21:45:02.208825   30286 kubelet.go:1705] >>__DBG_handlePodResourcesResize_PATCHING_APISERVER_DONE: pod=pod1
I0609 21:45:36.642618   30286 kubelet.go:1652] >>__DBG_canResizePod_ENTER: pod=pod2. Allocations: map[cpu:{i:{value:3 scale:0} d:{Dec:<nil>} s:3 Format:DecimalSI}]
I0609 21:45:36.642658   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod2. active_pod=kube-dns-747b6894bc-h8d95, Allocations: map[cpu:{i:{value:260 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} memory:{i:{value:115343360 scale:0} d:{Dec:<nil>} s: Format:BinarySI}]
I0609 21:45:36.642690   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod2. active_pod=pod1, Allocations: map[cpu:{i:{value:10 scale:0} d:{Dec:<nil>} s:10 Format:DecimalSI}]
I0609 21:45:36.642712   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod2. active_pod=pod2, Allocations: map[cpu:{i:{value:3 scale:0} d:{Dec:<nil>} s:3 Format:DecimalSI}]
I0609 21:45:36.644603   30286 kubelet.go:1665] >>__DBG_canResizePod_RESIZE_ADMIT_FAILED: pod=pod2. reason: OutOfcpu, message: Node didn't have enough resource: cpu, requested: 6000, used: 10260, capacity: 16000
root@fw0000358:~/jun01_vkInPlacePodVS_v1_19_dev/pkg/kubelet#
```

Reply:
About your scenario, yes in that case, kubectl user will see allocated > allocatable. I think that's ok - it would be a transient UX problem. 

But CRI update being held up may be an issue for the system... for that duration, we would be oversubscribed. At some point I did put a TODO marker in my code to possibly use a goroutine for apiserver patch op. Not sure why i dropped it.. i think i should add that back despite some added code complexity.

In either case, unless I missed something, we don't need to suffer the lock across PATCH call to apiserver. Concur?

Reply:
> If we can end up with allocated > allocatable, that means a kubelet restart will result in one of the pods being rejected, as it reads pods from the source to admit them.

good point. we have to patch holding the lock, ugh. i'll look into it after i'm done with resource quota and limit range, and will also fix the explicit update.

Reply:
Sounds good.  We have tests that test the throughput of pod startup.  If it has a significant negative impact, we can consider the approach you currently have, and document the potential issues.

Reply:
Commit https://github.com/vinaykul/kubernetes/pull/1/commits/4fae5883dd5a00a0914c0b468a965964f264b27a - Modified resource quota and limitranger plugins to handle updates to Requests and Limits, and modified kubelet to hold the lock across patch - it addresses the kubelet restart issue that you identified. Thanks!